### PR TITLE
Closes #43. Add Service Fabric client certificate id property

### LINKED ISSUE #43: Allow ServiceFabric backend clientCertificateId reference
@mirsaeedi One more haha...
We have a need to reference the Service Fabric client certificate using `clientCertificateId`. This property is available in [API version 2021-08-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.apimanagement/2021-08-01/service/backends?tabs=json).

The value in the generated ARM template would be a `resourceId`, i.e. 
```
"serviceFabricCluster": {
    "clientCertificateId": "[resourceId('Microsoft.ApiManagement/service/certificates', parameters('ApimServiceName'), '[clientcertId]')]",
     ...
}
```
I have working code I can submit a PR for, but we'd need to update the API schema version. Lemme know your thoughts :)

### PR DESCRIPTION
**Pull request by levimatheri**
Fixes #43

### CODE REVIEW

**Comment by levimatheri** [comment]
Please review @mirsaeedi

File: src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs [hunk 1]
```diff
@@ -2,7 +2,8 @@
 {
 	public class ServiceFabricCluster
 	{
-		public string ClientCertificatethumbprint { get; set; }
+        public string ClientCertificateId { get; set; }
+        public string ClientCertificatethumbprint { get; set; }
```
Comment (mirsaeedi) [inline src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs:6]: nit: Please fix indentation.

File: src/apimtemplate/Core/Mapping/BackendMapper.cs [hunk 2]
```diff
@@ -8,7 +9,12 @@ public class BackendMapper: IMapper
 	{
 		public void Map(IMapperConfigurationExpression cfg)
 		{
-			cfg.CreateMap<BackendDeploymentDefinition, BackendProperties>();
+			cfg.CreateMap<BackendDeploymentDefinition, BackendProperties>()
+				.ForMember(dst => dst.Properties, opt => opt.MapFrom(src => src.Properties))
+				.ForPath(dst => dst.Properties.ServiceFabricCluster.ClientCertificateId, 
+						opt => opt.MapFrom(src => string.IsNullOrEmpty(src.Properties.ServiceFabricCluster.ClientCertificateId)
```
Comment (mirsaeedi) [inline src/apimtemplate/Core/Mapping/BackendMapper.cs]: should we do `src.?Properties?.ServiceFabricCluster?.ClientCertificateId`?

File: src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs (hunk 1 above)
Comment (levimatheri) [reply src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs:6]: Indentation looks good to me, no?

File: src/apimtemplate/Core/Mapping/BackendMapper.cs (hunk 2 above)
Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: I get this error if I add the null-propagator 
image

Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: It seems that Automapper automatically handles the null propagation. Also see [answer from StackOverflow](https://stackoverflow.com/a/29234649/8519078). I tested removing some & all backend properties and it didn't throw any exceptions.

Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Actually I think we still need to check for null, I tested with [this yml](https://github.com/mirsaeedi/dotnet-apim/blob/master/src/Examples/Definitions/Backend.yml) and it's adding `serviceFabriccluster` property 
image
image

I'll see how I can fix this.

File: src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs (hunk 1 above)
Comment (mirsaeedi) [reply src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs:6]: It's how github shows it:
image

File: src/apimtemplate/Core/Mapping/BackendMapper.cs (hunk 2 above)
Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: @mirsaeedi I'm stuck here. I tried this, but it still adds `serviceFabricCluster` even if it's not specified in the yml.
```
cfg.CreateMap<BackendDeploymentDefinition, BackendProperties>()
				.ForMember(dst => dst.Properties, opt => opt.MapFrom(src => src.Properties ?? null))
				.ForPath(dst => dst.Properties.ServiceFabricCluster, opt =>
				{
					opt.Condition(src => src.SourceMember != null);
					opt.MapFrom(src => src.Properties.ServiceFabricCluster ?? null);
				})
				.ForPath(dst => dst.Properties.ServiceFabricCluster.ClientCertificateId, opt =>
				{
					opt.Condition(src => src.Source?.Properties?.ServiceFabricCluster != null && !string.IsNullOrEmpty(src.SourceMember));
					opt.MapFrom(src => $"[resourceId('{ResourceType.Certificate}', parameters('ApimServiceName'), '{src.Properties.ServiceFabricCluster.ClientCertificateId}')]");
				});
```

Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Okay I think I figured it out. Please review @mirsaeedi

Comment (mirsaeedi) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Hey bro. Probably we can give a try to `aftermap`:

```C#
cfg.CreateMap<BackendDeploymentDefinition, BackendProperties>()
				.AfterMap((backendDeploymentDefinition, backendProperties) =>
				{
					if (backendDeploymentDefinition?.Properties?.ServiceFabricCluster?.ClientCertificateId is not null)
					{
						backendProperties.Properties.ServiceFabricCluster.ClientCertificateId = $"[resourceId('{ResourceType.Certificate}', parameters('ApimServiceName'), '{backendDeploymentDefinition.Properties.ServiceFabricCluster.ClientCertificateId}')]";
					}
				});
```

Comment (mirsaeedi) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Hey @levimatheri, can we get this PR out soon. Have you seen my comment? thanks.

**Comment by levimatheri** [comment]
@mirsaeedi any update on this? I pushed some changes

File: src/apimtemplate/Core/Mapping/BackendMapper.cs (hunk 2 above)
Comment (mirsaeedi) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Hey @levimatheri, I commented 8 days ago but my comments were in the `Pending` status. Sorry for the confusion.

Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Okay no problem @mirsaeedi. I'll try your suggestion and submit soon.

File: src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs (hunk 1 above)
Comment (levimatheri) [reply src/apimtemplate/ApimEntities/Backend/ServiceFabricCluster.cs:6]: Fixed.

File: src/apimtemplate/Core/Mapping/BackendMapper.cs (hunk 2 above)
Comment (levimatheri) [reply src/apimtemplate/Core/Mapping/BackendMapper.cs]: Wow, you're awesome man, it worked!! :)

**Comment by levimatheri** [comment]
@mirsaeedi I've pushed your suggested fix which works. Thanks

**Review by mirsaeedi — APPROVED** [review]
Looks good to me. Will publish the change today. Thank you very much for your effort.

**Comment by mirsaeedi** [comment]
@levimatheri  the package is published.

**Pull request merged**